torch/testing/_internal/composite_compliance.py

```diff
@@ -410,3 +441,61 @@ def check_backward_formula(op, args, kwargs):
                     f"- wrapped_kwargs: {which_kwargs_are_wrapped}\n"
                     f"- wrapped_grads: {which_grad_is_batched}\n"
                 )
+
+# Checks if the forward AD formula is composite compliant by testing
+# all possible permutations of {primals, tangents} being
+# CompositeCompliantTensor or regular Tensors.
+def check_forward_ad_formula(op, args, kwargs):
+    assert op.supports_forward_ad
+
+    CCT = generate_cct(dispatch=True, alias=False)
+    # Permutations of arg and kwargs in CCT.
+    for choice in generate_subclass_choices_args_kwargs(args, kwargs, CCT):
+        new_args, new_kwargs, which_args_are_wrapped, which_kwargs_are_wrapped = choice
+
+        def maybe_tangent(t):
+            assert type(t) is not CCT
+            # Generate `tangent` tensor
+            # if given object is a Tensor and requires grad is set.
+            if isinstance(t, torch.Tensor) and t.requires_grad:
+                return torch.randn_like(t)
+            elif is_tensorlist(t):
+                return list(torch.randn_like(e) if e.requires_grad else None for e in t)
+            return None
+
+        tangent_args = tuple(maybe_tangent(arg) for arg in args)
+        flat_kwargs, spec = tree_flatten(kwargs)
+        flat_tangent_kwargs = tuple(maybe_tangent(arg) for arg in flat_kwargs)
+        tangent_kwargs = tree_unflatten(flat_tangent_kwargs, spec)
+
+        # Permutations tangent arg and tangent kwargs in CCT.
+        for tang_choice in generate_subclass_choices_args_kwargs(tangent_args, tangent_kwargs, CCT):
+            new_tang_args, new_tang_kwargs, \
+                which_tang_args_are_wrapped, which_tang_kwargs_are_wrapped = tang_choice
+
+            with fwAD.dual_level():
+                def maybe_make_dual(dual):
+                    # Returns dual tensor if primal is a tensor/tensor subclass
+                    # with requires_grad set.
+                    primal, tangent = dual
+                    if isinstance(primal, torch.Tensor) and primal.requires_grad:
+                        return fwAD.make_dual(primal, tangent)
```

Comment:
No need to do the branch, right? We've already checked during generation of the tangent. If that sounds correct then we should have a `if tangent is not None: assert primal.requires_grad` here.

Reply:
I think we need this. If `primal` is a Tensor without requires_grad then we want to take the fallback `return primal`, right?

Reply:
At this point `primal` could we `Tensor, List, List[Tensor] or any object`.

Reply:
Oh I see. Yes, you're right.